### ISSUE TITLE
Port twisted.web.http_headers to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6080]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6080|
|Type|enhancement|
|Created|2012-10-08 11:50:25Z|
|Branch|https://github.com/twisted/twisted/tree/http-headers-py3-6080|

This is a dependency of both the client and the server.

<details><summary>Searchable metadata</summary>

```
trac-id__6080 6080
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_http_headers_py3_6080 branches/http-headers-py3-6080
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1349697025000000 1349697025000000
changetime__1349823499000000 1349823499000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [36003]) Branching to 'http-headers-py3-6080'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Switched it to `MutableMapping` which wasn't _too_ bad, but it's not quite mechanical.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/http-headers-py3-6080).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

[#5829](https://github.com/twisted/twisted/issues/5829) was opened for `MutableMapping`, so that is either a duplicate or it should be referenced/closed when merging this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Looks good, with two minor fixes; merge when done:

  1. Add `twisted.web.test` to `_twistedpython3.py`.
  1. Some of the methods still refer to `C{str}` and strings instead of bytes, e.g. `hasHeader`, `removeHeader`, etc..

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [36040]) Address review comments

refs [#6080](https://github.com/twisted/twisted/issues/6080)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [36042]) Merge http-headers-py3-6080

Author: exarkun
Reviewer: itamarst
Fixes: [#6080](https://github.com/twisted/twisted/issues/6080)

Port `twisted.web.http_headers` to Python 3.  Also fix some structural
documentation issues.
